src/server/dcp_server/models.py

```diff
@@ -60,12 +75,475 @@ def masks_to_outlines(self, mask):
+
+
+    def get_center_of_mass(self, mask: np.ndarray) -> np.ndarray:
+        """
+        Compute the centers of mass for each object in a mask.
+
+        Args:
+            mask (np.ndarray): The input mask containing labeled objects.
+
+        Returns:
+            np.ndarray: An array of coordinates (row, column, channel) representing the centers of mass for each object.
+        """
+
+        # Compute the centers of mass for each labeled object in the mask
+        centers_of_mass = [
+        (int(x[0]), int(x[1]), int(x[2])) if mask.ndim >= 3 else (int(x[0]), int(x[1]), -1)
+        for x in center_of_mass(mask, mask, np.arange(1, mask.max() + 1))
+        ]
+
+        return centers_of_mass
+
+
+    def get_centered_patches(self, img, mask, p_size: int, noise_intensity=5):
+
+        ''' 
+        Extracts centered patches from the input image based on the centers of objects identified in the mask.
+
+        Args:
+            img: The input image.
+            mask: The mask representing the objects in the image.
+            p_size (int): The size of the patches to extract.
+            noise_intensity: The intensity of noise to add to the patches.
+
+        '''
+
+        patches, labels = [], []
+        # is_train = mask.ndim == 3
+
+        centers_of_mass = self.get_center_of_mass(mask)
+            # Crop patches around each center of mass and save them
+        for i, c in enumerate(centers_of_mass):
+            c_x, c_y, label = c
+
+            
+            # if is_train:
+            patch = self.pad_centered_padded_patch(img.copy(), (c_x, c_y), (p_size, p_size), mask=mask, noise_intensity=noise_intensity)
+            # else:
+            #     patch = self.pad_centered_padded_patch(img.copy(), (c_x, c_y), (p_size, p_size), mask=mask, noise_intensity=noise_intensity)
+            
+            patches.append(patch)
+            labels.append(label)
+
+        return patches, labels
+
+    def create_patch_dataset(self, imgs, masks, black_bg:bool, include_mask:bool, **kwargs):
+        '''
+        TODO: Split img and masks into patches of equal size which are centered around the cells.
+        The algorithm should first run through all images to find the max cell size, and use
+        the max cell size to define the patch size. All patches and masks should then be returned
+        in the same format as imgs and masks (same type, i.e. check if tensor or np.array and same 
+        convention of dims, e.g.  CxHxW)
+        Args:
+            imgs ():
+            masks ():
+            black_bg (bool): Flag indicating whether to use a black background for patches.
+            include_mask (bool): Flag indicating whether to include the mask along with patches.
+        '''
+
+        noise_intensity = kwargs.get("data", {}).get("noise_intensity", 5)
+
+        #max_patch_size = np.max([self.find_max_patch_size(mask) in masks])
+        max_patch_size = kwargs.get("data", {}).get("patch_size", 64)
```

Comment:
max_patch_size isn't called in this function. remove?

Reply:
okay I think it's best here to add an if statement, check if patch_size is given in the config and if not then call find_patch_size. Can you please implement this?

Reply:
Done